lib/block_view.go

```diff
@@ -7060,6 +7342,153 @@ func (bav *UtxoView) _connectSwapIdentity(
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyInvalidDerivedPublicKey, err.Error())
+	}
+
+	// Verify that the access signature is valid. This means the derived key is authorized.
+	err := _verifyAccessSignature(ownerPublicKey, derivedPublicKey,
+		txMeta.ExpirationBlock, txMeta.AccessSignature)
+	if err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyAccessSignatureNotValid, err.Error())
+	}
+
+	// Get current (previous) derived key entry. We might revert to it later so we copy it.
+	prevDerivedKeyEntry := bav._getDerivedKeyMappingForOwner(ownerPublicKey, derivedPublicKey)
+
+	// We make sure that the key hasn't been invalidated in the past. If that's the case,
+	// we return. A derived key can't be re-authorized after being revoked in the past.
+	if prevDerivedKeyEntry != nil && !prevDerivedKeyEntry.isDeleted {
+		if prevDerivedKeyEntry.OperationType == AuthorizeDerivedKeyOperationNotValid {
+			return 0, 0, nil, RuleErrorAuthorizeDerivedKeyDeletedDerivedPublicKey
+		}
+	}
+
+	// At this point we've verified the access signature, which means the derived key is authorized
+	// to sign on behalf of the owner. We now add a temporary derived key to utxoView with
+	// OperationType set to Valid so we pass txn signature verification in _connectBasicTransfer.
+	derivedKeyEntry := DerivedKeyEntry {
+		*NewPublicKey(ownerPublicKey),
+		*NewPublicKey(derivedPublicKey),
+		txMeta.ExpirationBlock,
+		AuthorizeDerivedKeyOperationValid,
+		false,
+	}
+	bav._setDerivedKeyMapping(&derivedKeyEntry)
+
+	// Call _connectBasicTransfer to verify txn signature.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		// Since we've failed, we revert the utxoView mapping to what it was previously.
+		// We're doing this manually because we haven't fully connected the txn yet.
+		bav._deleteDerivedKeyMapping(&derivedKeyEntry)
+		bav._setDerivedKeyMapping(prevDerivedKeyEntry)
+		return 0, 0, nil, errors.Wrapf(err, "_connectAuthorizeDerivedKey: ")
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		// Since we've failed, we revert the utxoView mapping to what it was previously.
+		// We're doing this manually because we haven't fully connected the txn yet.
+		bav._deleteDerivedKeyMapping(&derivedKeyEntry)
+		bav._setDerivedKeyMapping(prevDerivedKeyEntry)
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyRequiresNonZeroInput
+	}
+
+	// Now that we've verified this txn is good to go, we update the utxoView entry
+	// OperationType to what it's supposed to be.
+	derivedKeyEntry.OperationType = txMeta.OperationType
+	bav._setDerivedKeyMapping(&derivedKeyEntry)
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
```

Comment:
Nit: It's signed either by the top-level key OR the derived key. Let's make that distinction here.